File: 6th semester/Programación avanzada/Tarea2/onp.c
/*
  Ignacio Alvarado Reyes A01656149
  Pedro Rangel Palacios A01650939
*/
#include <stdio.h>
#include <string.h>

#define MAX_LENGTH 500
#define CASE_CODE strcat(stack, charAdd); stackLength++; strcpy(last, stack);

void stacc(char exp[][MAX_LENGTH], int length);

int main (void){
  int expressions;
  scanf("%d", &expressions);

  char ec[expressions][MAX_LENGTH];
  for(int i=0; i<expressions; i++) scanf("%s", ec[i]);

  stacc(ec, expressions);
  return 0;
}

void stacc(char exp[][MAX_LENGTH], int length) {
  for(int i=0; i<length; i++){
    char ans[MAX_LENGTH] = "";
    char stack[MAX_LENGTH] = "";
    char last[MAX_LENGTH]= "";
    int n = 0;
    int stackLength = 0;

    while(exp[i][n] != '\0') {
      char current = exp[i][n];
      char charAdd[] = {current, '\0'};

      if(current == '+' || current == '-' || current == '/' || current == '*' || current == '^') {
        strcat(stack, charAdd);
        stackLength++;
        strcpy(last, stack);
      }
      else if(current == ')') {
        strcat(ans, (char[2]) {stack[stackLength-1], '\0'});
        char temp[MAX_LENGTH] = "";
        strncpy(temp, stack, stackLength-1);
        stackLength--;
        strcpy(last, temp);
        strcpy(stack, last);
      }else {
        if(current != '(') strcat(ans, charAdd);
      }
      n++;
    }
    printf("%s\n", ans);
  }
}

File: 6th semester/Programación avanzada/Tarea2/onp.java
import java.util.Scanner;
public class onp {
    public static void main (String args[]){
        Scanner sc= new Scanner(System.in);
        int expressions = sc.nextInt();
        sc.nextLine();
        String[] ec= new String [expressions];
        for (int i=0; i<ec.length;i++){
            ec[i]= sc.nextLine();
        }

        stacc(ec);
        
    }

    public static void stacc (String[] exp){
        for(int j=0;j<exp.length; j++){
            String ans="";
            String stack="";
            String last= "";
            for(int i=0 ; i<exp[j].length(); i++){
                switch (exp[j].charAt(i)){
                    case '(': 
                        break;
                    case '+':
                        stack+= exp[j].charAt(i);
                        last= stack;
                        break;
                    case '-':
                        stack+= exp[j].charAt(i);
                        last= stack;
                        break;
                    case '/':
                        stack+= exp[j].charAt(i);
                        last= stack;
                        break;
                    case '*':
                        stack+= exp[j].charAt(i);
                        last= stack;
                        break;
                    case '^':
                        stack+= exp[j].charAt(i);
                        last= stack;
                        break;
                    case ')':
                        //System.out.println(stack.charAt(last)+" index:"+last);
                        ans+= stack.charAt(stack.length()-1);
                        System.out.println("ANS = "+ans);
                        last= stack.substring(0, stack.length()-1);
                        System.out.println("LAST = "+last);
                        stack=last;
                        System.out.println("STACK = "+stack);
                        break;
                    default:
                        ans+=exp[j].charAt(i);
                        break;
                }
            }
            System.out.println(ans);
        }
       
    }
}